File: android/app/src/main/java/binding/PermBinding.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2023 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package binding

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import channel.perm.PermOps
import channel.perm.PrivateDnsState
import channel.perm.PrivateDnsStateKind
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import repository.Repos
import service.BiometricService
import service.ConnectivityService
import service.ContextService
import service.Flavor
import service.FlutterService
import service.NotificationService
import service.SystemNavService
import service.VpnPermissionService
import utils.cause
import utils.FamilyOnboardingNotification
import utils.Intents
import utils.Logger
import utils.OnboardingNotification


object PermBinding : PermOps {
    val vpnProfileActivated = MutableStateFlow(false)
    private const val BLOKADA_SIX_PACKAGE = "org.blokada.sex"
    private const val BLOKADA_SIX_OWNER = "blokada6"
    private const val NO_PROTECTION_OWNER = "none"

    private val flutter by lazy { FlutterService }
    private val notification by lazy { NotificationService }
    private val permsRepo by lazy { Repos.perms }
    private val vpnPerms by lazy { VpnPermissionService }
    private val connectivity by lazy { ConnectivityService }
    private val context by lazy { ContextService }
    private val intents by lazy { Intents }
    private val scope = CoroutineScope(Dispatchers.Main)
    private val biometric by lazy { BiometricService }

    init {
        PermOps.setUp(flutter.engine.dartExecutor.binaryMessenger, this)
    }

    override fun getPrivateDnsState(callback: (Result<PrivateDnsState>) -> Unit) {
        val current = connectivity.privateDns
        val state = PrivateDnsState(
            kind = if (current.isNullOrEmpty()) {
                PrivateDnsStateKind.DISABLED
            } else {
                PrivateDnsStateKind.ENABLED
            },
            serverUrl = current
        )
        callback(Result.success(state))
    }

    override fun doSetPrivateDnsEnabled(
        tag: String,
        alias: String,
        callback: (Result<Unit>) -> Unit
    ) {
        // Cannot be done on Android
        callback(Result.success(Unit))
    }

    override fun doSetDns(tag: String, callback: (Result<Unit>) -> Unit) {
        // Cannot be done on Android
        callback(Result.success(Unit))
    }

    override fun doNotificationEnabled(callback: (Result<Boolean>) -> Unit) {
        // TODO: actual perms?
        callback(Result.success(false))
    }

    override fun doVpnEnabled(callback: (Result<Boolean>) -> Unit) {
        val enabled = vpnPerms.hasPermission()
        vpnProfileActivated.value = enabled
        callback(Result.success(enabled))
    }

    override fun getParentDeviceProtectionOwner(callback: (Result<String>) -> Unit) {
        val ctx = context.requireContext()
        val sixInstalled =
            ctx.packageManager.getLaunchIntentForPackage(BLOKADA_SIX_PACKAGE) != null
        if (!sixInstalled) {
            callback(Result.success(NO_PROTECTION_OWNER))
            return
        }

        val owner = if (activeVpnLooksLikeBlokadaSix(ctx) || privateDnsLooksLikeBlokadaSix()) {
            BLOKADA_SIX_OWNER
        } else {
            NO_PROTECTION_OWNER
        }
        callback(Result.success(owner))
    }

    private fun activeVpnLooksLikeBlokadaSix(ctx: Context): Boolean {
        val manager = ctx.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetwork = manager.activeNetwork ?: return false
        val caps = manager.getNetworkCapabilities(activeNetwork) ?: return false
        if (!caps.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) return false
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return false

        val packages = ctx.packageManager.getPackagesForUid(caps.ownerUid) ?: return false
        return packages.contains(BLOKADA_SIX_PACKAGE)
    }

    private fun privateDnsLooksLikeBlokadaSix(): Boolean {
        val host = connectivity.privateDns ?: return false
        if (!host.endsWith(".cloud.blokada.org")) return false
        return host.removeSuffix(".cloud.blokada.org").contains("-")
    }

    override fun doOpenSettings(callback: (Result<Unit>) -> Unit) {
        val n = if (Flavor.isFamily()) FamilyOnboardingNotification() else OnboardingNotification();
        notification.show(n)
        val ctx = context.requireContext()
        val intent = intents.createNetworkSettingsIntent(ctx)
        intents.openIntentActivity(ctx, intent)
        callback(Result.success(Unit))
    }

    override fun doAskNotificationPerms(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            permsRepo.maybeDisplayNotificationPermsDialog()
        }
    }

    override fun doAskVpnPerms(callback: (Result<Unit>) -> Unit) {
        vpnPerms.askPermission()
        callback(Result.success(Unit))
    }

    override fun doAuthenticate(callback: (Result<Boolean>) -> Unit) {
        scope.launch {
            if (biometric.isBiometricReady(context.requireContext())) {
                try {
                    biometric.auth(context.requireFragment()) // Will throw on bad auth
                } catch (ex: Exception) {
                    Logger.e("SettingsAccount", "Could not authenticate".cause(ex));
                    callback(Result.success(false))
                    return@launch
                }
            }

            callback(Result.success(true))
        }
    }

    override fun isRunningOnMac(callback: (Result<Boolean>) -> Unit) {
        // Android apps cannot run on Mac, always return false
        callback(Result.success(false))
    }
}

File: ios/App/Binding/Perm/PermBinding.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2023 Blocka AB. All rights reserved.
//
//  @author Kar
//

import Foundation
import Factory
import Combine
import UIKit

class PermBinding: PermOps {

    var vpnProfilePerms: AnyPublisher<Granted, Never> {
        writeVpnProfilePerms.compactMap { $0 }.removeDuplicates().eraseToAnyPublisher()
    }

    @Injected(\.flutter) private var flutter
    @Injected(\.common) private var notification
    
    private lazy var privateDns = Services.privateDns
    private lazy var netx = Services.netx
    private lazy var permsRepo = Repos.permsRepo
    private lazy var systemNav = Services.systemNav
    private lazy var account = ViewModels.account

    fileprivate let writeVpnProfilePerms = CurrentValueSubject<Granted?, Never>(nil)

    private var cancellables = Set<AnyCancellable>()

    init() {
        PermOpsSetup.setUp(binaryMessenger: flutter.getMessenger(), api: self)
        onVpnPerms()
    }
    
    func doAuthenticate(completion: @escaping (Result<Bool, any Error>) -> Void) {
        account.authenticate(ok: { _ in
            completion(.success(true))
        }, fail: { _ in
            completion(.success(false))
        })
    }

    func doSetPrivateDnsEnabled(tag: String, alias: String, completion: @escaping (Result<Void, Error>) -> Void) {
        privateDns.savePrivateDnsProfile(tag: tag, name: alias)
        .sink(
            onFailure: { err in completion(Result.failure(err))},
            onSuccess: { completion(Result.success(())) }
        )
        .store(in: &cancellables)
    }

    func doSetDns(tag: String, completion: @escaping (Result<Void, Error>) -> Void) {
        privateDns.savePrivateDnsProfile(tag: tag, name: nil)
        .sink(
            onFailure: { err in completion(Result.failure(err))},
            onSuccess: { completion(Result.success(())) }
        )
        .store(in: &cancellables)
    }

    func doNotificationEnabled(completion: @escaping (Result<Bool, Error>) -> Void) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            DispatchQueue.main.async {
                if (settings.authorizationStatus == .notDetermined) {
                    return completion(.success(false))
                }

                // Ignore other states, assuming we should not show the perm dialog (so return true)
                completion(.success(true))
            }
        }
    }

    func doVpnEnabled(completion: @escaping (Result<Granted, Error>) -> Void) {
        // If current value is non-nil, return immediately
        if let currentValue = writeVpnProfilePerms.value {
            completion(.success(currentValue))
            return
        }
        
        // Otherwise, wait for the next non-nil value
        writeVpnProfilePerms
            .compactMap { $0 } // Filter out nil values
            .first()           // Take the first non-nil value
            .sink(
                receiveCompletion: { result in
                    if case let .failure(error) = result {
                        completion(.failure(error))
                    }
                },
                receiveValue: { value in
                    completion(.success(value))
                }
            )
            .store(in: &cancellables)
    }

    func getParentDeviceProtectionOwner(completion: @escaping (Result<String, Error>) -> Void) {
        guard flutter.isFlavorFamily else {
            completion(.success(BlokadaSixProtectionOwnerMarker.ownerNone))
            return
        }

        guard canOpenBlokadaSix() else {
            completion(.success(BlokadaSixProtectionOwnerMarker.ownerNone))
            return
        }

        guard let storage = UserDefaults(
            suiteName: BlokadaSixProtectionOwnerMarker.storageSuite
        ) else {
            completion(.success(BlokadaSixProtectionOwnerMarker.ownerBlokadaSix))
            return
        }

        let owner = storage.string(forKey: BlokadaSixProtectionOwnerMarker.ownerKey)
        let updatedAt = storage.double(forKey: BlokadaSixProtectionOwnerMarker.updatedAtKey)
        let isFresh = BlokadaSixProtectionOwnerMarker.isFresh(updatedAt: updatedAt)

        if owner == BlokadaSixProtectionOwnerMarker.ownerNone && isFresh {
            completion(.success(BlokadaSixProtectionOwnerMarker.ownerNone))
            return
        }

        if owner == BlokadaSixProtectionOwnerMarker.ownerBlokadaSix && isFresh {
            completion(.success(BlokadaSixProtectionOwnerMarker.ownerBlokadaSix))
            return
        }

        completion(.success(BlokadaSixProtectionOwnerMarker.ownerBlokadaSix))
    }

    /// iOS does not expose another app's active DNS/VPN profile. If Blokada 6
    /// is installed but has not yet written a fresh marker, Family assumes the
    /// parent wants to keep managing this device in Blokada 6 and stays out.
    private func canOpenBlokadaSix() -> Bool {
        guard let url = URL(string: "six://") else { return false }
        return UIApplication.shared.canOpenURL(url)
    }

    func doOpenSettings(completion: @escaping (Result<Void, Error>) -> Void) {
        // Ask for notif perms oportunisticly (will help with onboarding)
        self.notification.askForPermissions()
            .receive(on: RunLoop.main)
            .sink(
                onFailure: { err in
                    self.openSystemSettings()
                    completion(Result.success(()))
                },
                onSuccess: {
                    self.openSystemSettings()
                    completion(Result.success(()))
                }
            )
            .store(in: &cancellables)
    }
    
    func doAskNotificationPerms(completion: @escaping (Result<Void, Error>) -> Void) {
        self.notification.askForPermissions()
            .receive(on: RunLoop.main)
            .sink()
            .store(in: &cancellables)
        completion(Result.success(()))
    }
    
    func doAskVpnPerms(completion: @escaping (Result<Void, Error>) -> Void) {
        self.permsRepo.askVpnProfilePerms()
            .receive(on: RunLoop.main)
            .sink(
                onFailure: { err in
                    completion(Result.failure(err))
                },
                onSuccess: {
                    completion(Result.success(()))
                }
            )
            .store(in: &cancellables)
    }

    func getPrivateDnsState(completion: @escaping (Result<PrivateDnsState, any Error>) -> Void) {
        privateDns.getPrivateDnsState()
        .sink(
            onValue: { value in
                let state = PrivateDnsState(
                    kind: self.mapPrivateDnsStateKind(value.kind),
                    serverUrl: value.serverUrl
                )
                completion(Result.success(state))
            },
            onFailure: { err in
                completion(Result.failure(err))
            }
        )
        .store(in: &cancellables)
    }

    func isRunningOnMac(completion: @escaping (Result<Bool, any Error>) -> Void) {
        // Check if this is an iPad app running on macOS
        if #available(iOS 14.0, *) {
            let isOnMac = ProcessInfo.processInfo.isiOSAppOnMac
            completion(Result.success(isOnMac))
        } else {
            // iOS versions before 14.0 cannot run on Mac
            completion(Result.success(false))
        }
    }

    func openSystemSettings() {
        let id = self.flutter.isFlavorFamily ? NOTIF_ONBOARDING_FAMILY : NOTIF_ONBOARDING
        self.notification.scheduleNotification(id: id, when: Date().addingTimeInterval(3))
        self.systemNav.openSystemSettings()
    }

    private func onVpnPerms() {
        netx.getPermsPublisher()
        .sink(onValue: { it in
            self.writeVpnProfilePerms.send(it)
        })
        .store(in: &cancellables)
    }

    private func mapPrivateDnsStateKind(_ kind: PrivateDnsProfileStateKind) -> PrivateDnsStateKind {
        switch kind {
        case .enabled:
            return .enabled
        case .disabled:
            return .disabled
        case .unavailable:
            return .unavailable
        }
    }
}

extension Container {
    var perm: Factory<PermBinding> {
        self { PermBinding() }.singleton
    }
}
